File: 05-memory-management/filemap.c
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <err.h>

#define BUFFER_SIZE	1000
#define ALLOC_SIZE	(100*1024*1024)

static char command[BUFFER_SIZE];
static char file_contents[BUFFER_SIZE];
static char overwrite_data[] = "HELLO";

int main(void)
{
	pid_t pid;

	pid = getpid();
	snprintf(command, BUFFER_SIZE, "cat /proc/%d/maps", pid);

	puts("*** memory map before mapping file ***");
	fflush(stdout);
	system(command);

	int fd;
	fd = open("testfile", O_RDWR);
	if (fd == -1)
		err(EXIT_FAILURE, "open() failed");

	char * file_contents;
	file_contents = mmap(NULL, ALLOC_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
	if (file_contents == (void *) -1) {
		warn("mmap() failed");
		goto close_file;
	}

	puts("");
	printf("*** succeeded to map file: address = %p; size = 0x%x ***\n",
		file_contents, ALLOC_SIZE);

	puts("");
	puts("*** memory map after mapping file ***");
	fflush(stdout);
	system(command);

	puts("");
	printf("*** file contens before overwrite mapped region: %s", file_contents);

	// overwrite mapped region
	memcpy(file_contents, overwrite_data, strlen(overwrite_data));

	puts("");
	printf("*** overwritten mapped region with: %s\n", file_contents);

	if (munmap(file_contents, ALLOC_SIZE) == -1)
		warn("munmap() failed");
close_file:
	if (close(fd) == -1)
		warn("close() failed");
	exit(EXIT_SUCCESS);
}

File: 05-memory-management/misc/filemap.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"strconv"
	"syscall"
)

func main() {
	pid := os.Getpid()
	fmt.Println("*** testfileのメモリマップ前のプロセスの仮想アドレス空間 ***")
	command := exec.Command("cat", "/proc/"+strconv.Itoa(pid)+"/maps")
	command.Stdout = os.Stdout
	err := command.Run()
	if err != nil {
		log.Fatal("catの実行に失敗しました")
	}

	file, err := os.OpenFile("testfile", os.O_RDWR, 0)
	if err != nil {
		log.Fatal("testfileを開けませんでした")
	}
	defer file.Close()

	// mmap()システムコールの呼び出しによって1GBのメモリ領域を獲得
	data, err := syscall.Mmap(int(file.Fd()), 0, 5, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
	if err != nil {
		log.Fatal("mmap()に失敗しました")
	}

	fmt.Println("")
	fmt.Printf("testfileをマップしたアドレス: 0x%p\n", &data[0])
	fmt.Println("")

	fmt.Println("*** testfileのメモリマップ後のプロセスの仮想アドレス空間 ***")
	command = exec.Command("cat", "/proc/"+strconv.Itoa(pid)+"/maps")
	command.Stdout = os.Stdout
	err = command.Run()
	if err != nil {
		log.Fatal("catの実行に失敗しました")
	}

	// マップしたファイルの中身を書き換える
	replaceBytes := []byte("HELLO")
	for i, _ := range data {
		data[i] = replaceBytes[i]
	}
}
